actionpack/test/dispatch/test_response_test.rb

```diff
@@ -38,4 +38,32 @@ def assert_response_code_range(range, predicate)
     assert_kind_of(Nokogiri::XML::Document, response.parsed_body)
     assert_equal(response.parsed_body.at_xpath("/html/body/div").text, "Content")
   end
+
+  if RUBY_VERSION >= "3.0"
+    test "JSON response pattern matching" do
+      response = ActionDispatch::TestResponse.create(200, { "Content-Type" => "application/json" }, '{ "foo": "fighters" }')
+
+      # rubocop:disable Lint/Syntax
+      assert_pattern { response.parsed_body.with_indifferent_access => { foo: /fighter/ } }
```

Comment:
Why the `with_indifferent_access`?

Reply:
@rafaelfranca I've come to understand that the JSON integration with Ruby's pattern matching is more complicated than I thought, so I opened https://github.com/rails/rails/pull/49003 to incorporate it into the test suite. I'll simplify this PR (https://github.com/rails/rails/pull/49001) so that it only mentions HTML pattern matching.